File: backend/lib/wiki_cache.py
import asyncio
import time
from pathlib import Path

from backend.config import WIKI_DATA_DIR
from backend.lib.frontmatter import parse_frontmatter
from backend.lib.doc_stats import compute_stats

_cached_docs: list[dict] | None = None
_cache_timestamp: float = 0
_CACHE_TTL = 5.0


def invalidate_wiki_cache() -> None:
    global _cached_docs, _cache_timestamp
    _cached_docs = None
    _cache_timestamp = 0


async def get_all_docs() -> list[dict]:
    global _cached_docs, _cache_timestamp
    now = time.time()
    if _cached_docs is not None and now - _cache_timestamp < _CACHE_TTL:
        return _cached_docs

    docs: list[dict] = []
    await _walk_dir(WIKI_DATA_DIR, "", docs)
    _cached_docs = docs
    _cache_timestamp = time.time()
    return docs


async def _walk_dir(dir_path: Path, relative_path: str, collector: list[dict]) -> None:
    try:
        entries = await asyncio.to_thread(lambda: list(dir_path.iterdir()))
    except OSError:
        return

    md_files: list[dict] = []
    subdirs: list[tuple[Path, str]] = []

    for entry in entries:
        if entry.name.startswith("."):
            continue
        rel_path = f"{relative_path}/{entry.name}" if relative_path else entry.name

        if entry.is_dir():
            subdirs.append((entry, rel_path))
        elif entry.name.endswith(".md"):
            md_files.append({
                "full_path": entry,
                "rel_path": rel_path,
                "name": entry.name,
                "is_index": entry.name == "_index.md",
            })

    for sub_path, sub_rel in subdirs:
        await _walk_dir(sub_path, sub_rel, collector)

    async def _read_one(f: dict) -> dict | None:
        try:
            content = await asyncio.to_thread(f["full_path"].read_text, "utf-8")
            stat = await asyncio.to_thread(f["full_path"].stat)
            return {**f, "content": content, "mtime": stat.st_mtime * 1000}
        except OSError:
            return None

    results = await asyncio.gather(*[_read_one(f) for f in md_files])

    for r in results:
        if r is None:
            continue
        import re
        display_path = re.sub(r"\.md$", "", r["rel_path"])
        display_path = re.sub(r"/_index$", "", display_path)

        title_match = re.search(r"^#\s+(.+)$", r["content"], re.MULTILINE)
        title = title_match.group(1).strip() if title_match else display_path.split("/")[-1]

        parsed = parse_frontmatter(r["content"])
        stats = compute_stats(r["content"])

        collector.append({
            "path": display_path,
            "title": title,
            "content": r["content"],
            "wordCount": stats["word_count"],
            "tags": parsed["frontmatter"].get("tags", []),
            "isFolder": r["is_index"],
            "mtime": r["mtime"],
        })

File: lib/wiki-cache.ts
import fs from 'fs/promises';
import path from 'path';
import { parseFrontmatter } from '@/lib/frontmatter';
import { computeStats } from '@/lib/doc-stats';

const WIKI_DATA_DIR = path.join(process.cwd(), 'wiki-data');

/* ── 类型 ── */

export interface DocEntry {
  path: string;         // 去掉 .md 的相对路径
  title: string;
  content: string;
  wordCount: number;
  tags: string[];
  isFolder: boolean;
  mtime: number;        // 文件修改时间戳
}

/* ── 缓存核心 ── */

let cachedDocs: DocEntry[] | null = null;
let cacheTimestamp = 0;
const CACHE_TTL_MS = 5_000; // 5 秒 TTL，写操作后手动失效

/** 手动使缓存失效（写入/删除/重命名后调用） */
export function invalidateWikiCache(): void {
  cachedDocs = null;
  cacheTimestamp = 0;
}

/** 获取全量文档列表（带缓存） */
export async function getAllDocs(): Promise<DocEntry[]> {
  const now = Date.now();
  if (cachedDocs && now - cacheTimestamp < CACHE_TTL_MS) {
    return cachedDocs;
  }

  const docs: DocEntry[] = [];
  await walkDir(WIKI_DATA_DIR, '', docs);
  cachedDocs = docs;
  cacheTimestamp = Date.now();
  return docs;
}

/* ── 目录遍历（内部） ── */

async function walkDir(dirPath: string, relativePath: string, collector: DocEntry[]): Promise<void> {
  let entries;
  try {
    entries = await fs.readdir(dirPath, { withFileTypes: true });
  } catch {
    return;
  }

  // 批量收集需要 stat 的文件
  const mdFiles: { fullPath: string; relPath: string; name: string; isIndex: boolean }[] = [];

  for (const entry of entries) {
    if (entry.name.startsWith('.')) continue;
    const fullPath = path.join(dirPath, entry.name);
    const relPath = relativePath ? `${relativePath}/${entry.name}` : entry.name;

    if (entry.isDirectory()) {
      await walkDir(fullPath, relPath, collector);
    } else if (entry.name.endsWith('.md')) {
      mdFiles.push({ fullPath, relPath, name: entry.name, isIndex: entry.name === '_index.md' });
    }
  }

  // 并行读取文件内容和 stat
  const results = await Promise.all(
    mdFiles.map(async (f) => {
      try {
        const [content, stat] = await Promise.all([
          fs.readFile(f.fullPath, 'utf-8'),
          fs.stat(f.fullPath),
        ]);
        return { ...f, content, mtime: stat.mtimeMs };
      } catch {
        return null;
      }
    }),
  );

  for (const r of results) {
    if (!r) continue;
    const displayPath = r.relPath.replace(/\.md$/, '').replace(/\/_index$/, '');
    const titleMatch = r.content.match(/^#\s+(.+)$/m);
    const title = titleMatch ? titleMatch[1].trim() : displayPath.split('/').pop() || '';
    const { frontmatter } = parseFrontmatter(r.content);
    const stats = computeStats(r.content);

    collector.push({
      path: displayPath,
      title,
      content: r.content,
      wordCount: stats.wordCount,
      tags: frontmatter.tags || [],
      isFolder: r.isIndex,
      mtime: r.mtime,
    });
  }
}
